tests/zfs-tests/cmd/libzfs_input_check/libzfs_input_check.c

```diff
@@ -0,0 +1,896 @@
+	nvlist_t *required = fnvlist_alloc();
+
+	fnvlist_add_string(required, bookmark, snapshot);
+
+	IOC_INPUT_TEST_WILD(ZFS_IOC_BOOKMARK, pool, required, NULL, 0);
+
+	nvlist_free(required);
+}
+
+static void
+test_get_bookmarks(const char *dataset)
+{
+	nvlist_t *optional = fnvlist_alloc();
+
+	fnvlist_add_boolean(optional, "guid");
+	fnvlist_add_boolean(optional, "createtxg");
+	fnvlist_add_boolean(optional, "creation");
+
+	IOC_INPUT_TEST_WILD(ZFS_IOC_GET_BOOKMARKS, dataset, NULL, optional, 0);
+
+	nvlist_free(optional);
+}
+
+static void
+test_destroy_bookmarks(const char *pool, const char *bookmark)
+{
+	nvlist_t *required = fnvlist_alloc();
+
+	fnvlist_add_boolean(required, bookmark);
+
+	IOC_INPUT_TEST_WILD(ZFS_IOC_DESTROY_BOOKMARKS, pool, required, NULL, 0);
+
+	nvlist_free(required);
+}
+
+static void
+test_clone(const char *snapshot, const char *clone)
+{
+	nvlist_t *required = fnvlist_alloc();
+	nvlist_t *optional = fnvlist_alloc();
+	nvlist_t *props = fnvlist_alloc();
+
+	fnvlist_add_string(required, "origin", snapshot);
+
+	IOC_INPUT_TEST(ZFS_IOC_CLONE, clone, required, NULL, 0);
+
+	nvlist_free(props);
+	nvlist_free(optional);
+	nvlist_free(required);
+}
+
+static void
+test_rollback(const char *dataset, const char *snapshot)
+{
+	nvlist_t *optional = fnvlist_alloc();
+
+	fnvlist_add_string(optional, "target", snapshot);
+
+	IOC_INPUT_TEST(ZFS_IOC_ROLLBACK, dataset, NULL, optional, B_FALSE);
+
+	nvlist_free(optional);
+}
+
+static void
+test_hold(const char *pool, const char *snapshot)
+{
+	nvlist_t *required = fnvlist_alloc();
+	nvlist_t *optional = fnvlist_alloc();
+	nvlist_t *holds = fnvlist_alloc();
+
+	fnvlist_add_string(holds, snapshot, "libzfs_check_hold");
+	fnvlist_add_nvlist(required, "holds", holds);
+	fnvlist_add_int32(optional, "cleanup_fd", zfs_fd);
+
+	IOC_INPUT_TEST(ZFS_IOC_HOLD, pool, required, optional, 0);
+
+	nvlist_free(holds);
+	nvlist_free(optional);
+	nvlist_free(required);
+}
+
+static void
+test_get_holds(const char *snapshot)
+{
+	IOC_INPUT_TEST(ZFS_IOC_GET_HOLDS, snapshot, NULL, NULL, 0);
+}
+
+static void
+test_release(const char *pool, const char *snapshot)
+{
+	nvlist_t *required = fnvlist_alloc();
+	nvlist_t *release = fnvlist_alloc();
+
+	fnvlist_add_boolean(release, "libzfs_check_hold");
+	fnvlist_add_nvlist(required, snapshot, release);
+
+	IOC_INPUT_TEST_WILD(ZFS_IOC_RELEASE, pool, required, NULL, 0);
+
+	nvlist_free(release);
+	nvlist_free(required);
+}
+
+
+static void
+test_send_new(const char *snapshot, int fd)
+{
+	nvlist_t *required = fnvlist_alloc();
+	nvlist_t *optional = fnvlist_alloc();
+
+	fnvlist_add_int32(required, "fd", fd);
+
+	fnvlist_add_boolean(optional, "largeblockok");
+	fnvlist_add_boolean(optional, "embedok");
+	fnvlist_add_boolean(optional, "compressok");
+	fnvlist_add_boolean(optional, "rawok");
+#if 0
+	fnvlist_add_string(optional, "fromsnap", from);
+	fnvlist_add_uint64(optional, "resume_object", resumeobj);
+	fnvlist_add_uint64(optional, "resume_offset", offset);
+#endif
```

Comment:
Why are these disabled?  I presume because the required test case would be more complicated.  That's OK but then let's add a comment.

Reply:
comment added